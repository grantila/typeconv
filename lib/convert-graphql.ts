import {
	convertCoreTypesToGraphql,
	convertGraphqlToCoreTypes,
	CoreTypesToGraphqlOptions,
	GraphqlToCoreTypesOptions,
} from 'core-types-graphql'

import { Reader } from "./reader"
import { Writer } from "./writer"
import { userPackage, userPackageUrl } from "./package"


export function getGraphQLReader( graphqlOptions?: GraphqlToCoreTypesOptions )
: Reader
{
	return {
		kind: 'gql',
		read( source, options )
		{
			return convertGraphqlToCoreTypes(
				source,
				{
					...options,
					...graphqlOptions,
				}
			);
		}
	};
}

export function getGraphQLWriter( graphqlOptions?: CoreTypesToGraphqlOptions )
: Writer
{
	return {
		kind: 'gql',
		write( doc, options )
		{
			return convertCoreTypesToGraphql(
				doc,
				{
					warn: options.warn,
					filename: options.filename,
					sourceFilename: options.sourceFilename,
					userPackage,
					userPackageUrl,
					...graphqlOptions,
				}
			);
		}
	};
}
