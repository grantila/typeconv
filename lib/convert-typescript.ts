import {
	convertCoreTypesToTypeScript,
	convertTypeScriptToCoreTypes,
	FromTsOptions,
	ToTsOptions,
} from 'core-types-ts'

import { Reader } from "./reader"
import { Writer } from "./writer"
import { userPackage, userPackageUrl } from "./package"


export function getTypeScriptReader( tsOptions?: FromTsOptions ): Reader
{
	return {
		kind: 'ts',
		read( source, { warn, filename } )
		{
			return convertTypeScriptToCoreTypes(
				source,
				{
					warn,
					...tsOptions,
				}
			);
		}
	};
}

export function getTypeScriptWriter( tsOptions?: ToTsOptions ): Writer
{
	return {
		kind: 'ts',
		write( doc, { warn, filename, sourceFilename } )
		{
			return convertCoreTypesToTypeScript(
				doc,
				{
					warn,
					filename,
					sourceFilename,
					userPackage,
					userPackageUrl,
					...tsOptions,
				}
			);
		}
	};
}
