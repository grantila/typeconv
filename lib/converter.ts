import {
	decorateError,
	decorateErrorMeta,
	isCoreTypesError,
	NamedType,
	NodeDocument,
	simplify,
	WarnFunction,
	ConversionResult as CoreTypesConversionResult,
} from 'core-types'

import type { Reader, ReaderOptions } from './reader'
import type { Shortcut, Writer, WriterOptions } from './writer'
import type { SourceÂ } from './file'
import { getSource, writeFile, relFile } from './file'
import { formatError } from './error'


export interface Target
{
	filename: string;
	relFilename?: string;
}

export interface ConversionResultInformation
{
	in: Omit< CoreTypesConversionResult, 'data' >;
	out: Omit< CoreTypesConversionResult, 'data' >;
}

export type ConversionResult< T > = ConversionResultInformation & { data: T };

export interface Converter
{
	/**
	 * Convert and save to file
	 */
	convert( from: Source, to: Target ): Promise< ConversionResult< void > >;

	/**
	 * Convert and return result as string
	 */
	convert( from: Source ): Promise< ConversionResult< string > >;

	/**
	 * The current working directory for this converter
	 */
	cwd: string;
}

export type ConvertMapFunction =
	( node: NamedType, index: number, array: ReadonlyArray< NamedType > ) =>
		NamedType;
export type ConvertFilterFunction =
	( node: NamedType, index: number, array: ReadonlyArray< NamedType > ) =>
		boolean;
export type ConvertTransformFunction = ( doc: NodeDocument ) => NodeDocument;

export interface ConvertOptions
{
	/**
	 * The current working directory to use when converting to or from files.
	 * Is not necessary for string-to-string conversion.
	 */
	cwd?: string;

	/**
	 * When simplify is true, the converter will let core-types _compress_ the
	 * types after having converted from {reader} format to core-types.
	 * This is usually recommended, but may cause some annotations (comments)
	 * to be dropped.
	 *
	 * @default
	 * 	true
	 */
	simplify?: boolean;

	/**
	 * Custom map function for transforming each type after it has been
	 * converted *from* the source type (and after it has been simplified),
	 * but before it's written to the target type system.
	 *
	 * If `filter` is used as well, this runs before `filter`.
	 * If `transform` is used as well, this runs before `transform`.
	 */
	map?: ConvertMapFunction;

	/**
	 * Custom filter function for filtering types after they have been
	 * converted *from* the source type.
	 *
	 * If `map` is used as well, this runs after `map`.
	 * If `transform` is used as well, this runs before `transform`.
	 */
	filter?: ConvertFilterFunction;

	/**
	 * Custom filter function for filtering types after they have been
	 * converted *from* the source type.
	 *
	 * If `map` is used as well, this runs after `map`.
	 * If `filter` is used as well, this runs after `filter`.
	 */
	transform?: ConvertTransformFunction;

	/**
	 * Shortcut reader and writer if possible (bypassing core-types).
	 *
	 * @default
	 *  true
	 */
	shortcut?: boolean;
}

export function makeConverter(
	reader: Reader,
	writer: Writer,
	options?: ConvertOptions
)
: Converter
{
	const { shortcut = true, cwd = process.cwd( ) } = options ?? { };

	const relFilename = ( filename: string ) =>
		relFile( options?.cwd, filename );

	async function convert( from: Source, to?: Target )
	: Promise< ConversionResult< void | string > >
	{
		const { data, filename } = await getSource( from );

		const warn: WarnFunction = ( msg, meta ) =>
		{
			const fullMeta = decorateErrorMeta(
				{ ...meta },
				{ filename, source: data }
			);
			console.warn( formatError( msg, fullMeta ) );
		}

		const toFilename =
			to?.relFilename ?? relFilename( to?.filename ?? '' );

		const readOpts: ReaderOptions = {
			warn,
			filename: filename ? relFilename( filename ) : undefined,
		};

		const writeOpts: WriterOptions = {
			warn,
			...(
				to
				? { filename: toFilename }
				: { }
			),
			...(
				filename
				? { sourceFilename: relFilename( filename ) }
				: { }
			),
			rawInput: data,
		};

		const convertShortcut = async ( ) =>
		{
			const convertedTypes: Array< string > = [ ];
			const notConvertedTypes: Array< string > = [ ];

			const shortcut = writer.shortcut?.[ reader.kind ] as Shortcut;
			const {
				data: output,
				convertedTypes: outConvertedTypes,
				notConvertedTypes: outNotConvertedTypes,
			} = await shortcut( data, readOpts, writeOpts, reader );

			return {
				output,
				convertedTypes,
				notConvertedTypes,
				outConvertedTypes,
				outNotConvertedTypes,
			};
		};

		const convertDefault = async ( ) =>
		{
			const { data: doc, convertedTypes, notConvertedTypes } =
				await reader.read( data, readOpts );

			const simplifiedDoc =
				options?.simplify === false
				? doc
				: simplify( doc );

			const { map, filter, transform } = options ?? { };

			const mappedDoc =
				typeof map === 'function'
				? {
					...simplifiedDoc,
					types: doc.types.map( ( type, index ) =>
						map( type, index, doc.types )
					),
				}
				: simplifiedDoc;

			const filteredDoc =
				typeof filter === 'function'
				? {
					...mappedDoc,
					types: doc.types.filter( ( type, index ) =>
						filter( type, index, doc.types )
					),
				}
				: mappedDoc;

			const finalDoc =
				typeof transform === 'function'
				? transform( filteredDoc )
				: filteredDoc;

			const {
				data: output,
				convertedTypes: outConvertedTypes,
				notConvertedTypes: outNotConvertedTypes
			} = await writer.write( finalDoc, writeOpts );

			return {
				output,
				convertedTypes,
				notConvertedTypes,
				outConvertedTypes,
				outNotConvertedTypes,
			};
		};

		try
		{
			const {
				output,
				convertedTypes,
				notConvertedTypes,
				outConvertedTypes,
				outNotConvertedTypes,
			} =
				( shortcut && writer.shortcut?.[ reader.kind ] )
				? await convertShortcut( )
				: await convertDefault( );

			const info: ConversionResultInformation = {
				in: { convertedTypes, notConvertedTypes },
				out: {
					convertedTypes: outConvertedTypes,
					notConvertedTypes: outNotConvertedTypes,
				},
			}

			if ( typeof to?.filename === 'undefined' )
				return { data: output, ...info } as ConversionResult< string >;

			// Only write non-empty files
			if ( outConvertedTypes.length > 0 )
				await writeFile( to?.filename, output );

			return info as ConversionResult< void >;
		}
		catch ( err )
		{
			if ( isCoreTypesError( err ) )
				decorateError( err, {
					source: data,
					...( filename ? { filename } : { } ),
				} );
			throw err;
		}
	}

	return { convert: convert as Converter[ 'convert' ], cwd };
}
