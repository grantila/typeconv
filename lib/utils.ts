
export function stringify( value: any )
{
	return JSON.stringify( value, null, 2 );
}
