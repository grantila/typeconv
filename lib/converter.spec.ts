import { makeConverter } from './converter'
import {
	getJsonSchemaReader,
	getJsonSchemaWriter,
	getOpenApiWriter,
} from './convert-json-schema'
import { getTypeScriptWriter } from './convert-typescript'
import { getGraphQLWriter } from './convert-graphql'
import { withConsoleWarn } from '../test/utils'

describe( "converter", ( ) =>
{
	it( "self-reading and writing JSON Schema", async ( ) =>
	{
		const { convert } = makeConverter(
			getJsonSchemaReader( ),
			getJsonSchemaWriter( ),
		);
		const example = {
			definitions: {
				foo: { type: 'string' }
			}
		};
		const { data: out } =
			await convert( { data: JSON.stringify( example, null, 4 ) } );

		expect( JSON.parse( out ) ).toMatchObject( example );
	} );

	describe( "json-schema-to-ts", ( ) =>
	{
		it( "self-reading and writing JSON Schema", async ( ) =>
		{
			const { convert } = makeConverter(
				getJsonSchemaReader( ),
				getTypeScriptWriter( {
					noDescriptiveHeader: true,
					noDisableLintHeader: true,
				} ),
			);
			const example = {
				definitions: {
					foo: { type: 'string' }
				}
			};
			const { data: out } =
				await convert( { data: JSON.stringify( example, null, 4 ) } );

			expect( out ).toMatchSnapshot( );
		} );
	} );

	describe( "warn", ( ) =>
	{
		it( "should invoke warn function when present", withConsoleWarn(
			async ( consoleWarn ) =>
		{
			const warn = jest.fn( );
			const { convert } = makeConverter(
				getJsonSchemaReader( ),
				getGraphQLWriter( { warn } ),
			);
			const example = {
				definitions: {
					foo: { type: 'null' }
				}
			};
			await convert( { data: JSON.stringify( example, null, 4 ) } );

			expect( warn ).toHaveBeenCalledWith(
				"Type 'null' not supported",
				expect.any( Error )
			);
			expect( consoleWarn ).toHaveBeenCalledTimes( 0 );
		} ) );

		it( "should invoke console.warn by default", withConsoleWarn(
			async ( consoleWarn ) =>
		{
			const { convert } = makeConverter(
				getJsonSchemaReader( ),
				getGraphQLWriter( ),
			);
			const example = {
				definitions: {
					foo: { type: 'null' }
				}
			};
			await convert( { data: JSON.stringify( example, null, 4 ) } );

			expect( consoleWarn ).toHaveBeenCalledWith(
				"Type 'null' not supported"
			);
		} ) );
	} );

	describe( "shortcut", ( ) =>
	{
		it( "should run shortcut if possible", async ( ) =>
		{
			const { convert } = makeConverter(
				getJsonSchemaReader( ),
				getOpenApiWriter( {
					format: 'yaml',
					title: 'Title',
					version: '1',
					schemaVersion: '3.0.0',
				} ),
				{ shortcut: true }
			);
			const example = {
				definitions: {
					foo: { type: 'string' }
				}
			};
			const { data: out } =
				await convert( { data: JSON.stringify( example, null, 4 ) } );

			expect( out ).toMatchSnapshot( );
		} );
	} );
} );
