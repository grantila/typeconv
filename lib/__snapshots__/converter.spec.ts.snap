// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converter json-schema-to-ts self-reading and writing JSON Schema 1`] = `
"export type foo = string;
"
`;

exports[`converter shortcut should run shortcut if possible 1`] = `
"openapi: 3.0.0
info:
  title: Title
  version: '1'
paths: {}
components:
  schemas:
    foo:
      type: string
"
`;
